File: Java/PathFlow/src/main/java/com/eaglerobotics/app/A_Star.java
package com.eaglerobotics.app;
import java.lang.Math;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

/**
 * Hello world!
 *
 */


public class A_Star 
{
    private List<Node> grid = new ArrayList<Node>();
    private Double[] object_size;
    private Double[] grid_size;

    public A_Star(Double[] _object_size, Double[] _grid_size) {
        object_size = _object_size;
        grid_size = _grid_size;

        for (double i=object_size[0]; i<grid_size[0]; i+=object_size[0]) {
            for (double j=object_size[1]; j<grid_size[1]; j+=object_size[1]) {
                // System.out.printf("%s, %s\n", Math.round(1000.0*i)/1000.0, Math.round(1000.0*j)/1000.0);
                grid.add(new Node(round(i), round(j), false));
            }
        }
    }

    private Node findPointOnGrid(Node node) {
        for (Node g : grid) {
            if (Arrays.equals(g.getLocation(), node.getLocation())) {
                return g;
            }
        }
        return null;
    }

    private double round(double x) {
        return Math.round(1000.0*x)/1000.0;
    }

    private List<Node> getNeighbors(Node point) {
        List<Double[]> neighbors = new ArrayList<Double[]>();
        List<Node> final_neighbors = new ArrayList<Node>();
        Double[] location = point.getLocation();
        // System.out.printf("Computing Neighbors for (%s, %s)\n", location[0], location[1]);
        neighbors.add(new Double[] {round(location[0]+object_size[0]), round(location[1])});
        neighbors.add(new Double[] {round(location[0]-object_size[0]), round(location[1])});
        neighbors.add(new Double[] {round(location[0]), round(location[1]+object_size[1])});
        neighbors.add(new Double[] {round(location[0]), round(location[1]-object_size[1])});
        neighbors.add(new Double[] {round(location[0]-object_size[0]), round(location[1]-object_size[1])});
        neighbors.add(new Double[] {round(location[0]+object_size[0]), round(location[1]+object_size[1])});
        for (Double[] n : neighbors) {
            Node g = findPointOnGrid(new Node(n[0], n[1], false));
            if (g != null && g.obstacle == false) {
                // System.out.printf("Found neighbor (%s %s)\n", g.getLocation()[0], g.getLocation()[1]);
                final_neighbors.add(g);
            }
        }
        return final_neighbors;
    }

    private double getDistance(Node pointA, Node pointB) {
        Double[] coordA = pointA.getLocation();
        Double[] coordB = pointB.getLocation();
        double xDist = Math.abs(coordA[0] - coordB[0]);
        double yDist = Math.abs(coordA[1] - coordB[1]);
        return 14 * Math.min(xDist, yDist) + 10 * Math.abs(xDist - yDist);
    }

    public List<Node> compute(Node startNode, Node endNode) throws RuntimeException {
        startNode = findPointOnGrid(startNode.toNearestGrid(object_size));
        endNode = findPointOnGrid(endNode.toNearestGrid(object_size));
        System.out.println(Arrays.deepToString(startNode.getLocation()));
        System.out.println(Arrays.deepToString(endNode.getLocation()));
        if (startNode == null || startNode.obstacle) {
            throw new RuntimeException("Invalid Start Node!!!");
        }
        if (endNode.obstacle || endNode == null) {
            throw new RuntimeException("Invalid End Node!!!");
        }
        startNode.set_g_cost(0.0);
        List<Node> open = new ArrayList<Node>();
        open.add(startNode);
        List<Node> closed = new ArrayList<Node>();

        while (open.size() > 0) {
            Node currentNode = open.get(0);
            for (int i=1; i<open.size(); i++) {
                if (open.get(i).get_f_cost() < currentNode.get_f_cost() || (open.get(i).get_f_cost() == currentNode.get_f_cost() && open.get(i).get_h_cost() < currentNode.get_h_cost())) {
                    currentNode = open.get(i);
                }
            }
            open.remove(currentNode);
            closed.add(currentNode);

            // System.out.println(Arrays.deepToString(currentNode.getLocation()));

            if (currentNode == endNode) {
                Node lastNode = endNode;
                List<Node> chain = new ArrayList<Node>();
                while (lastNode.parent != null) {
                    chain.add(lastNode);
                    // System.out.println(Arrays.deepToString(lastNode.getLocation()));
                    lastNode = lastNode.parent;
                }
                chain.add(startNode);
                chain = chain.reversed();
                return chain;
            }

            for (Node n : getNeighbors(currentNode)) {
                if (closed.contains(n)) {
                    continue;
                }
                double newMoveToNeighborCost = currentNode.get_g_cost() + getDistance(currentNode, n);
                if (newMoveToNeighborCost < n.get_g_cost() || !open.contains(n)) {
                    n.set_g_cost(newMoveToNeighborCost);
                    n.set_h_cost(getDistance(n, endNode));
                    n.set_parent(currentNode);
                    if (!open.contains(n)) {
                        open.add(n);
                    }
                }
            }
        }
        // return new ArrayList<Node>();
        throw new RuntimeException("No path found");
    }

    public Double[][] nodeListToDouble(List<Node> path) {
        Double[][] finalList = new Double[path.size()][2];
        for (int i=0; i<path.size(); i++) {
            Double[] point = {path.get(i).getLocation()[0], path.get(i).getLocation()[1]};
            finalList[i] = point;
        }
        return finalList;
    }

    public void rectangularObstacle(Double[] topLeft, Double[] bottomRight) {
        for (Node g : grid) {
            Double[] gCoord = g.getLocation();
            if (gCoord[0] >= topLeft[0] && gCoord[0] <= bottomRight[0]) {
                if (gCoord[1] >= bottomRight[1] && gCoord[1] <= topLeft[1]) {
                    g.obstacle = true;
                    // System.out.printf("Obstacle at (%s, %s)\n", g.getLocation()[0], g.getLocation()[1]);
                }
            }
        }
    }

    public static void main( String[] args )
    {
        System.out.println( "Hello World from A*!" );
        Double[] o = {0.66,0.66};
        Double[] g = {8.2,16.0};
        A_Star a = new A_Star(o, g);
        // System.out.println(a.getNeighbors(new Node(1, 1, false)));
        // System.out.println(a.getDistance(new Node(1, 1, false), new Node(6, 3, false)));
        // a.rectangularObstacle(new Double[] {2.0,3.0}, new Double[] {4.0,3.0});
        System.out.println(Arrays.deepToString(a.nodeListToDouble(a.compute(new Node(1,1,false), new Node(2,4,false)))));
    }
}

File: Python/PathFlow/pathflow/pathfinding/a_star.py
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import matplotlib.animation as animation
import warnings

from pathflow.motionplanning import Bezier

class A_Star:
    def __init__(self, object_size: tuple[int, int], grid_size: tuple[int, int]) -> None:
        warnings.warn("The Python implementation does not yet support floating point grids or locations!")
        self.object_size = object_size
        self.grid_size = grid_size
        self.grid = []
        for i in range(object_size[0], grid_size[0], object_size[0]):
            for j in range(object_size[1], grid_size[1], object_size[1]):
                self.grid.append(Point([i, j], False))
        self.fig, self.ax1 = plt.subplots()
        self.obstacleRects = []
    def findPointOnGrid(self, point):
        for g in self.grid:
            if g.location == point.location:
                return g
        return None
    def _drawGrid(self):
        rect = patches.Rectangle((0, 0), self.grid_size[0], self.grid_size[1], linewidth=1, edgecolor='blue', facecolor='none')
        self.ax1.add_patch(rect)
        for g in self.grid:
            if g.obstacle:
                plt.plot(g.location[0], g.location[1], 'o', color='r')
            else:
                plt.plot(g.location[0], g.location[1], 'o', color='b')
    def _getNeighbors(self, point):
        neighbors = []
        location = point.location
        neighbors.append([location[0]+self.object_size[0], location[1]])
        neighbors.append([location[0]-self.object_size[0], location[1]])
        neighbors.append([location[0], location[1]+self.object_size[1]])
        neighbors.append([location[0], location[1]-self.object_size[1]])
        neighbors.append([location[0]-self.object_size[0], location[1]-self.object_size[1]])
        neighbors.append([location[0]+self.object_size[0], location[1]+self.object_size[1]])
        for n in neighbors:
            if self.findPointOnGrid(Point(n)) != None and self.findPointOnGrid(Point(n)).obstacle == False:
                yield self.findPointOnGrid(Point(n))
    def _getDistance(self, pointA, pointB):
        coordA = pointA.location
        coordB = pointB.location
        xDist = abs(coordA[0]-coordB[0])
        yDist = abs(coordA[1]-coordB[1])
        return 14*min(xDist, yDist) + 10*abs(yDist-xDist)
 
    def visualize(self, path, steps):
        b = Bezier()
        func = b.compute_bezier(path)
        trailx = []
        traily = []
        def animate(i):
            #print(b(i/100))
            self.ax1.clear()
            for r in self.obstacleRects:
                self.ax1.add_patch(r)
            self._drawGrid()
            self.ax1.plot(func(i/steps)[0], func(i/steps)[1], 'o', color='purple')
            trailx.append(func(i/steps)[0])
            traily.append(func(i/steps)[1])
            self.ax1.plot(trailx, traily, '--', color='black')
            self.ax1.plot(list(zip(*path))[0], list(zip(*path))[1], 'o', color='green')
        ani = animation.FuncAnimation(self.fig, animate, interval=1, frames=steps, repeat=False)
        # writervideo = animation.FFMpegWriter(fps=60) 
        # ani.save('bezier.mp4', writer=writervideo) 
        #plt.show()
        plt.draw()
        plt.waitforbuttonpress(0) # this will wait for indefinite time
        plt.close(self.fig)

    def rectangularObstacle(self, topLeft: list[float, float], bottomRight: list[float, float]):
        rect = patches.Rectangle(topLeft, bottomRight[0]-topLeft[0], bottomRight[1]-topLeft[1], linewidth=1, edgecolor='red', facecolor='red')
        self.obstacleRects.append(rect)
        for g in self.grid:
            gCoord = g.location
            print(gCoord)
            if gCoord[0] >= topLeft[0] and gCoord[0] <= bottomRight[0]:
                if gCoord[1] >= bottomRight[1] and gCoord[1] <= topLeft[1]:
                    g.obstacle = True



    def compute(self, start_point, end_point):
        start_point = self.findPointOnGrid(Point(start_point))
        end_point = self.findPointOnGrid(Point(end_point))
        for g in self.grid:
            g.fcost = 0
            g.gcost = 0
            g.hcost = 0
        start_point.gcost = 0
        open = [start_point]
        closed = []

        count = 0

        while len(open) > 0:
            current_node = open[0]
            for i in range(1, len(open)):
                if (open[i].fcost < current_node.fcost or (open[i].fcost == current_node.fcost and open[i].hcost < current_node.hcost)):
                    current_node = open[i]
            open.remove(current_node)
            closed.append(current_node)

            if (current_node == end_point):
                path = [end_point.location]
                currentNode = end_point
                while currentNode.parent != None:
                    path.append(currentNode.parent.location)
                    currentNode = currentNode.parent

                return list(reversed(path))
            for n in list(self._getNeighbors(current_node)):
                if n in closed:
                    continue
                newMoveCostToNeighbor = current_node.gcost + self._getDistance(current_node, n)
                if n.gcost == None or newMoveCostToNeighbor < n.gcost or n not in open:
                    n.gcost = newMoveCostToNeighbor
                    n.hcost = self._getDistance(n, end_point)
                    n.fcost = n.gcost + n.hcost
                    n.parent = current_node
                    if n not in open:
                        open.append(n)









# object_size = (1,1)
# grid_size = (5,5)
# grid = []



class Point:
    def __init__(self, location, obstacle=False) -> None:
        self.location = location
        self.obstacle = obstacle
        self.gcost = None
        self.hcost = None
        self.fcost = None
        self.parent = None
        pass



#obstacles = [[2,3], [3,3], [4,3]]
#for o in obstacles:
#    findPointOnGrid(grid, Point(o)).obstacle = True





#    print(open)




#end = a_star(findPointOnGrid(grid, Point([1,1])), findPointOnGrid(grid, Point([2,4])))
#path = [end.location]
#currentNode = end
#while currentNode.parent != None:
#    print(currentNode.parent.location)
#    path.append(currentNode.parent.location)
#    currentNode = currentNode.parent

#path = list(reversed(path))
#xs = list(zip(*path))[0]
#ys = list(zip(*path))[1]

#import numpy as np
#import matplotlib.animation as animation
#import math
#def compute_bezier(points):
#    points = np.array(points)
#    n = len(points) - 1
#    def bezier(t):
#        # print(t)
#        pose = np.array([0.0, 0.0])
#        if t<=0:
#            return points[0]
#        elif t>=1:
#            return points[-1]
#        else:
#            for i in range(n+1):
#                print(i)
#                # print(n, i)
#                pose += math.comb(n, i)*np.power(1-t, n-i)*np.power(t, i)*points[i]
#        return pose  
#    return bezier

#b = compute_bezier(path)
## print(b(0.31415))
#def animate(i):
#    #print(b(i/100))
#    ax.clear()
#    drawGrid(grid)
#    plt.plot(xs, ys, '-')
#    ax.plot(b(i/100)[0], b(i/100)[1], 'o')
#    ax.plot(list(zip(*path))[0], list(zip(*path))[1], 'o')
#    # ax1.plot(xar,yar)
#ani = animation.FuncAnimation(fig, animate, interval=1, frames=100)
#writervideo = animation.FFMpegWriter(fps=60) 
#ani.save('bezier.mp4', writer=writervideo) 
#plt.show()





#plt.show()

